File: 1. Implement Queue using Stacks.py
## Problem 1: (https://leetcode.com/problems/implement-queue-using-stacks/)

## both self.inStack and outStack are such that newest element is at end, ensuring O(1) during pop
## Always push into inStack, when we pop or peek we transer all elements from inStack to outStack, incase outStack is empty.
## If we pop and push alternatively it will not matter since the elements in "front" will get removed first from outStack.

class MyQueue:

    def __init__(self):
        self.inStack = []
        self.outStack = []
    def push(self, x: int) -> None:
        self.inStack.append(x)

    def pop(self) -> int:
        if self.outStack:
            return self.outStack.pop()
        else:
            while self.inStack:
                self.outStack.append(self.inStack.pop())
            return self.outStack.pop()

    def peek(self) -> int:
        if self.outStack:
            return self.outStack[-1]
        else:
            while self.inStack:
                self.outStack.append(self.inStack.pop())
            return self.outStack[-1]


        while self.outStack:
            self.inStack.append(self.outStack.pop())
        return peeked


    def empty(self) -> bool:
        return not self.inStack and not self.outStack



# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: 1. Implement Queue using Stacks.java
// Time Complexity : O(1) - Amortized time complexity (Worst case is O(N) but the worst case is rare here, on average it will be O(1).
// Space Complexity : O(N) space complexity where N is the number of elements (we are using 2 stacks though)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Getting used to Java Syntax, confusing between True(Python) and true(Java). Messing up semicolons etc


// Your code here along with comments explaining your approach
// The idea is to use 2 stacks approach. When there is a pop operation, check if the outStack is empty and push all elements from in stack to out stack. Simply pop the top element.
// When push is given, simply push the element into in Stack
// when peek is given, empty in stack into out stak and then peek at the out stack to get the first element in queue
// for isEmpty operation simply check to see if both stacks are empty, then true else false.

class MyQueue {
    Stack<Integer> inSt;
    Stack<Integer> outSt;

    public MyQueue() {
        this.inSt = new Stack<>();
        this.outSt = new Stack<>();
    }
    
    public void push(int x) {
        inSt.push(x);
    }
    
    public int pop() {
        if (outSt.isEmpty()){
            while(!inSt.isEmpty()){
                outSt.push(inSt.pop());
            }

        }
        return outSt.pop();
    }
    
    public int peek() { 
        if (outSt.isEmpty()){
            while(!inSt.isEmpty()){
                outSt.push(inSt.pop());
            }
        }
        return outSt.peek();
    }
    
    public boolean empty() {
        if (inSt.isEmpty() && outSt.isEmpty()){
            return true;
        }
        return false;
        
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */
